### PR TITLE
[Feature] Workflow Composer

### PR DESCRIPTION
**Pull request by shreyasbhat0**
# Pull Request Template

## Description

Workflow composer is responsible  for composing and deploying the necessary workflows.

### Components  

* Workflow Provider
* Workflow Config File
* Macros
* Operators
* Rust Boilerplate code

To create a workflow, the only necessary input is the `Workflow Config file`. The composition, direction of flow and actions are laid out in the Workflow Config file. This `Workflow Config file` is parsed by `Workflow Provider` from Tackle box, which in turn creates rust package for workflow with predefined `Rust Boilerplate code` , `Macros` , `Operator` and can be deployed to necessary environment for execution . 

Please select the options that are relevant.

- [ ] Bug fix (non-breaking change which fixes an issue)
- [x] New feature (non-breaking change which adds functionality)
- [ ] Breaking change (fix or feature that would cause existing functionality to not work as expected)
- [ ] This change requires a documentation update

## Checklist:

- [x] Pull Request follow only a single responsibility.
- [x] Code follows the style guidelines of this project
- [x] Performed a self-review of my own code
- [x] Commented my code, particularly in hard-to-understand areas
- [x] Made corresponding changes to the documentation
- [x] Changes generate no new warnings
- [ ] Added tests that prove fix is effective or the feature works
- [ ] New and existing unit tests pass locally with the changes
- [ ] Any dependent changes have been merged and published in downstream modules
- [x] Checked the code and corrected any misspellings

### CODE REVIEW

File: workflow/openwhisk_macro/src/lib.rs [hunk 1]
```diff
@@ -0,0 +1,79 @@
...
+        _ => false,
+    };
+
+    let auth_token = property_map["AuthKey"].clone();
+    let api_host = property_map["ApiHost"].clone();
+    let insecure = insecure;
+    let namespace = property_map["Namespace"].clone();
+
+    let impl_whisk_client = quote! {
+
```
Comment (MuhammedIrfan) [inline workflow/openwhisk_macro/src/lib.rs:53]: Remove line break

Comment (MuhammedIrfan) [inline workflow/openwhisk_macro/src/lib.rs]: Remove line break

File: workflow/openwhisk_macro/src/lib.rs [hunk 2]
```diff
@@ -0,0 +1,79 @@
...
+        impl #name {
+
```
Comment (MuhammedIrfan) [inline workflow/openwhisk_macro/src/lib.rs]: Remove line break

File: workflow/openwhisk_macro/src/lib.rs [hunk 3]
```diff
@@ -0,0 +1,79 @@
...
+            pub fn openwhisk_client(&self) -> OpenwhiskClient<NativeClient>{
+
```
Comment (MuhammedIrfan) [inline workflow/openwhisk_macro/src/lib.rs]: Remove line break

File: workflow/openwhisk_macro/src/lib.rs [hunk 4]
```diff
@@ -0,0 +1,79 @@
...
+                let wskprops  = WskProperties::new(#auth_token.to_string(), #api_host.to_string(), #insecure, #namespace.to_string());
+
+                OpenwhiskClient::<NativeClient>::new(Some(&wskprops))
+            }
+
+            fn run(&mut self){
+
```
Comment (MuhammedIrfan) [inline workflow/openwhisk_macro/src/lib.rs]: Remove line break

File: workflow/openwhisk_macro/src/lib.rs [hunk 5]
```diff
@@ -0,0 +1,79 @@
...
+                    let result = self.openwhisk_client()
+                        .actions()
+                        .invoke(&self.action_name, serde_json::to_value(self.input.clone()).unwrap(), true, true).unwrap();
+                    self.output = serde_json::from_value(result).unwrap();
+
```
Comment (MuhammedIrfan) [inline workflow/openwhisk_macro/src/lib.rs]: Remove line break

File: workflow/openwhisk_macro/src/lib.rs [hunk 6]
```diff
@@ -0,0 +1,79 @@
...
+                let wskprops  = WskProperties::new(#auth_token.to_string(), #api_host.to_string(), #insecure, #namespace.to_string());
+
+                OpenwhiskClient::<NativeClient>::new(Some(&wskprops))
+            }
+
+            fn run(&mut self){
+
+                    let result = self.openwhisk_client()
+                        .actions()
+                        .invoke(&self.action_name, serde_json::to_value(self.input.clone()).unwrap(), true, true).unwrap();
```
Comment (MuhammedIrfan) [inline workflow/openwhisk_macro/src/lib.rs]: add line break

File: workflow/openwhisk_macro/src/lib.rs [hunk 7]
```diff
@@ -0,0 +1,79 @@
...
+                    self.output = serde_json::from_value(result).unwrap();
+
+            }
+
+            fn get_action_name(&self) -> String {
+
```
Comment (MuhammedIrfan) [inline workflow/openwhisk_macro/src/lib.rs]: Remove line break

File: workflow/openwhisk_macro/src/lib.rs [hunk 8]
```diff
@@ -0,0 +1,79 @@
...
+            fn run(&mut self){
+
+                    let result = self.openwhisk_client()
+                        .actions()
+                        .invoke(&self.action_name, serde_json::to_value(self.input.clone()).unwrap(), true, true).unwrap();
+                    self.output = serde_json::from_value(result).unwrap();
+
+            }
+
+            fn get_action_name(&self) -> String {
```
Comment (MuhammedIrfan) [inline workflow/openwhisk_macro/src/lib.rs]: get_action_name to action_name as return reference &str

File: workflow/openwhisk_macro/src/lib.rs [hunk 9]
```diff
@@ -0,0 +1,79 @@
...
+                
+                self.action_name.clone()
+            }
+        }
+    };
```
Comment (MuhammedIrfan) [inline workflow/openwhisk_macro/src/lib.rs:75]: Add line break

File: workflow/openwhisk_macro/src/lib.rs [hunk 10]
```diff
@@ -0,0 +1,79 @@
...
+            syn::Meta::NameValue(syn::MetaNameValue {
+                path,
+                lit: syn::Lit::Str(s),
+                ..
+            }) => (path, s.value()),
+            _ => (syn::Path::into(attribute.path), "".to_string()),
+        };
+
+        for seg in path.segments {
+            property_map.insert(seg.ident.to_string().to_string(), value.clone());
```
Comment (MuhammedIrfan) [inline workflow/openwhisk_macro/src/lib.rs]: remove .to_string()

File: workflow/openwhisk_macro/src/lib.rs [hunk 11]
```diff
@@ -0,0 +1,79 @@
...
+        impl #name {
+
+            pub fn openwhisk_client(&self) -> OpenwhiskClient<NativeClient>{
+
+                let wskprops  = WskProperties::new(#auth_token.to_string(), #api_host.to_string(), #insecure, #namespace.to_string());
+
+                OpenwhiskClient::<NativeClient>::new(Some(&wskprops))
+            }
+
+            fn run(&mut self){
```
Comment (MuhammedIrfan) [inline workflow/openwhisk_macro/src/lib.rs]: Return as result to handle error outside

File: workflow/openwhisk_macro/src/lib.rs (hunk 6 above)
Comment (MuhammedIrfan) [inline workflow/openwhisk_macro/src/lib.rs]: Convert to recoverable errors

File: workflow/openwhisk_macro/src/lib.rs [hunk 12]
```diff
@@ -0,0 +1,79 @@
...
+
+                    let result = self.openwhisk_client()
+                        .actions()
+                        .invoke(&self.action_name, serde_json::to_value(self.input.clone()).unwrap(), true, true).unwrap();
+                    self.output = serde_json::from_value(result).unwrap();
```
Comment (MuhammedIrfan) [inline workflow/openwhisk_macro/src/lib.rs]: Convert to recoverable errors

File: workflow/providers/hooks/constants.py [hunk 13]
```diff
@@ -0,0 +1,188 @@
...
+}}
+
+impl Workflow {{
+    pub fn new<T: 'static + Execute>(task: T) -> Self {{
+        Self {{
+            vertex: Box::new(task),
+            edge: None,
+        }}
+    }}
+    pub fn init(&mut self, flow: Option<Box<dyn FlowExecutor>>) -> &mut Self {{
```
Comment (MuhammedIrfan) [inline workflow/providers/hooks/constants.py:30]: Add line break

File: workflow/providers/hooks/functions.py [hunk 14]
```diff
@@ -0,0 +1,418 @@
...
+    current_task_type_map = ""
+    map_task_name = ""
+
+    local_pipe_dic = dict()
+    local_pipe_list = []
+
+    for flow in flow_list:
+
+        if flow['type'] == "Init":            
+            for key, values in task_store_copy[flow['task_name']].items():
```
Comment (MuhammedIrfan) [inline workflow/providers/hooks/functions.py:179]: key:value to field: type

File: workflow/providers/hooks/functions.py [hunk 15]
```diff
@@ -0,0 +1,418 @@
...
+                            depend_task_type_map = i['type']
+
+                if depend_task_type_map != None:
+                    dep_type = re.findall(r'\<.*?\>', depend_task_type_map)
+                    depend_task_type_map = "".join(dep_type)
+
+            task_struct_impl += f"""
+#[derive(Default, Debug, Clone, Serialize, Deserialize)]
+pub struct {map_task_name }MapOutput{{
+    model_prize_list : HashMap<{depend_task_type_map},{current_task_type_map}>
```
Comment (MuhammedIrfan) [inline workflow/providers/hooks/functions.py:248]: Change to field_name

**Review by MuhammedIrfan — CHANGES_REQUESTED** [review]

File: workflow/openwhisk_macro/src/lib.rs (hunk 9 above)
Comment (shreyasbhat0) [reply workflow/openwhisk_macro/src/lib.rs:75]: Fixed

File: workflow/openwhisk_macro/src/lib.rs (hunk 12 above)
Comment (shreyasbhat0) [reply workflow/openwhisk_macro/src/lib.rs]: Handled error to recoverable

File: workflow/providers/hooks/constants.py (hunk 13 above)
Comment (shreyasbhat0) [reply workflow/providers/hooks/constants.py:30]: Fixed line break

File: workflow/openwhisk_macro/src/lib.rs (hunk 6 above)
Comment (shreyasbhat0) [reply workflow/openwhisk_macro/src/lib.rs]: Fixed

File: workflow/openwhisk_macro/src/lib.rs (hunk 5 above)
Comment (shreyasbhat0) [reply workflow/openwhisk_macro/src/lib.rs]: Fixed

File: workflow/openwhisk_macro/src/lib.rs (hunk 4 above)
Comment (shreyasbhat0) [reply workflow/openwhisk_macro/src/lib.rs]: Fixed

File: workflow/openwhisk_macro/src/lib.rs (hunk 3 above)
Comment (shreyasbhat0) [reply workflow/openwhisk_macro/src/lib.rs]: fixed

File: workflow/openwhisk_macro/src/lib.rs (hunk 2 above)
Comment (shreyasbhat0) [reply workflow/openwhisk_macro/src/lib.rs]: fixed

File: workflow/openwhisk_macro/src/lib.rs (hunk 8 above)
Comment (shreyasbhat0) [reply workflow/openwhisk_macro/src/lib.rs]: fixed

File: workflow/openwhisk_macro/src/lib.rs (hunk 7 above)
Comment (shreyasbhat0) [reply workflow/openwhisk_macro/src/lib.rs]: fixed

File: workflow/openwhisk_macro/src/lib.rs (hunk 6 above)
Comment (shreyasbhat0) [reply workflow/openwhisk_macro/src/lib.rs]: fixed

File: workflow/openwhisk_macro/src/lib.rs (hunk 11 above)
Comment (shreyasbhat0) [reply workflow/openwhisk_macro/src/lib.rs]: fixed

File: workflow/openwhisk_macro/src/lib.rs (hunk 10 above)
Comment (shreyasbhat0) [reply workflow/openwhisk_macro/src/lib.rs]: fixed

File: workflow/providers/hooks/functions.py (hunk 14 above)
Comment (shreyasbhat0) [reply workflow/providers/hooks/functions.py:179]: fixed

File: workflow/providers/hooks/functions.py (hunk 15 above)
Comment (shreyasbhat0) [reply workflow/providers/hooks/functions.py:248]: fixed Changes mentioned

**Review by MuhammedIrfan — APPROVED** [review]

File: workflow/openwhisk_macro/src/lib.rs [hunk 16]
```diff
@@ -0,0 +1,77 @@
...
+                
+                self.output = serde_json::from_value(result).map_err(|e|e.to_string())?;
+                Ok(())
+            }
+
+            fn action_name(&self) -> &str {
+                self.action_name.as_str()
+            }
+        }
+    };
```
Comment (MuhammedIrfan) [inline workflow/openwhisk_macro/src/lib.rs:75]: Add a line break

**Pull request merged**